addons/pos_loyalty/static/src/overrides/components/payment_screen/payment_screen.js

```diff
@@ -41,24 +41,29 @@ patch(PaymentScreen.prototype, {
                     [[], pointChanges, newCodes]
                 );
                 // Payload may contain the points of the concerned coupons to be updated in case of error. (So that rewards can be corrected)
-                const { couponCache } = this.pos;
                 if (payload && payload.updated_points) {
                     for (const pointChange of Object.entries(payload.updated_points)) {
-                        if (couponCache[pointChange[0]]) {
-                            couponCache[pointChange[0]].balance = pointChange[1];
+                        const coupon = this.pos.models["loyalty.card"].get(pointChange[0]);
+                        if (coupon) {
+                            coupon.points = pointChange[1];
                         }
                     }
                 }
                 if (payload && payload.removed_coupons) {
                     for (const couponId of payload.removed_coupons) {
-                        if (couponCache[couponId]) {
-                            delete couponCache[couponId];
-                        }
+                        const coupon = this.pos.models["loyalty.card"].get(couponId);
+                        coupon && coupon.delete();
                     }
-                    this.currentOrder.uiState.codeActivatedCoupons =
-                        this.currentOrder.uiState.codeActivatedCoupons.filter(
-                            (coupon) => !payload.removed_coupons.includes(coupon.id)
-                        );
+                    this.currentOrder.update({
+                        code_activated_coupon_ids: [
+                            [
+                                "unlink",
+                                ...payload.removed_coupons.map((id) =>
+                                    this.pos.models["loyalty.card"].get(id)
+                                ),
+                            ],
+                        ],
+                    });
```

Comment:
Is this needed ? Normally delete() already disconnect all recrods ?

Reply:
Good catch!